zapcore/sampler.go

```diff
@@ -126,9 +193,32 @@ func (s *sampler) Check(ent Entry, ce *CheckedEntry) *CheckedEntry {
 	}
 
 	counter := s.counts.get(ent.Level, ent.Message)
-	n := counter.IncCheckReset(ent.Time, s.tick)
+	n, prevN := counter.IncCheckReset(ent.Time, s.tick)
 	if n > s.first && (n-s.first)%s.thereafter != 0 {
 		return ce
 	}
+	if s.reporting.enabled && prevN > 0 {
```

Comment:
rather than complicating the logic here, what if we just ran a function (similar to a callback) to report, and we passed in parameters to it?

separate the responsibility of how to form the message/log away from here.